File: src/main/java/com/bn/milk/criteria/BaseCriteria.java
package com.bn.milk.criteria;

import com.bn.milk.model.Driver;
import com.bn.milk.model.Order;
import com.bn.milk.model.Vehicle;

public abstract class BaseCriteria {
    protected double weight;
    protected String name;

    public BaseCriteria(String name, double weight) {
        this.name = name;
        this.weight = weight;
    }

    public abstract double calculate(Order order, Vehicle vehicle, Driver driver);

    public double getWeight() { return weight; }
    public void setWeight(double weight) { this.weight = weight; }
    public String getName() { return name; }
}

File: backend/src/criteria/BaseCriteria.ts
import { OrderDTO } from "../domains/dtos/orderDTO";
import { VehicleDTO } from "../domains/dtos/vehicleDTO";
import { DriverDTO } from "../domains/dtos/driverDTO";

export abstract class BaseCriteria {
  protected name: string;
  protected weight: number;

  constructor(name: string, weight: number) {
    this.name = name;
    this.weight = weight;
  }

  abstract calculate(
    order: OrderDTO,
    vehicle: VehicleDTO,
    driver: DriverDTO
  ): number;

  getName(): string {
    return this.name;
  }

  public setWeight(newWeight: number): void {
    this.weight = newWeight;
  }

  public getWeight(): number {
    return this.weight;
  }
}
